packages/@aws-cdk/aws-lambda/lib/function.ts

```diff
@@ -395,6 +409,41 @@ export class Function extends FunctionBase {
     for (const event of props.events || []) {
       this.addEventSource(event);
     }
+
+    // Log retention
+    if (props.logRetentionDays) {
+      // Custom resource provider
+      const provider = new SingletonFunction(this, 'LogRetentionProvider', {
+        code: Code.asset(path.join(__dirname, 'log-retention')),
+        runtime: Runtime.NodeJS810,
+        handler: 'index.handler',
+        uuid: 'aae0aa3c-5b4d-4f87-b02d-85b201efdd8a',
+        lambdaPurpose: 'LogRetention',
+      });
+
+      if (!Function.logRetentionRolePolicy) { // Avoid duplicate statements
+        provider.addToRolePolicy(
+          new iam.PolicyStatement()
+            .addActions('logs:PutRetentionPolicy', 'logs:DeleteRetentionPolicy')
+            // We need '*' here because we will also put a retention policy on
+            // the log group of the provider function. Referencing it's name
+            // creates a CF circular dependency.
+            .addAllResources()
+        );
+        Function.logRetentionRolePolicy = true;
+      }
+
+      // Need to use a CfnResource here to prevent lerna dependency cycles
+      // @aws-cdk/aws-cloudformation -> @aws-cdk/aws-lambda -> @aws-cdk/aws-cloudformation
+      new cdk.CfnResource(this, 'LogRetentionCustomResource', {
+        type: 'AWS::CloudFormation::CustomResource',
```

Comment:
Use a custom type name that indicates what this type is doing (I believe "AWS::Custom::XXX" or something along those lines)